src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -25,10 +27,10 @@ public sealed class TypeDescriptor
         // lock on it for thread safety. It is used from nearly
         // every call to this class, so it will be created soon after
         // class load anyway.
-        private static readonly WeakHashtable s_providerTable = new WeakHashtable();     // mapping of type or object hash to a provider list
-        private static readonly Hashtable s_providerTypeTable = new Hashtable();         // A direct mapping from type to provider.
-        private static readonly Hashtable s_defaultProviders = new Hashtable(); // A table of type -> default provider to track DefaultTypeDescriptionProviderAttributes.
-        private static WeakHashtable? s_associationTable;
+        private static readonly ConditionalWeakTable<object, TypeDescriptionNode?> s_providerTable = new ConditionalWeakTable<object, TypeDescriptionNode?>(); // mapping of type or object hash to a provider list
```

Comment:
WeakHashtable has different semantics than ConditionalWeakTable, in particular around when things are dropped from the collection.  ConditionalWeakTable uses weak references for all the keys, which means something added could be removed on the very next GC; in contrast, WeakHashtable had its own policy for when keys would be scavenged based on memory thresholds.  I don't think we can just change one to the other without a comprehensive performance analysis.